evm/src/generation/memory.rs

```diff
@@ -0,0 +1,52 @@
+use plonky2::field::types::Field;
+
+use crate::memory::memory_stark::MemoryOp;
+use crate::memory::segments::NUM_SEGMENTS;
+use crate::memory::VALUE_LIMBS;
+
+#[allow(unused)] // TODO: Should be used soon.
+#[derive(Debug)]
+pub(crate) struct MemoryState<F: Field> {
+    /// A log of each memory operation, in the order that it occurred.
+    pub log: Vec<MemoryOp<F>>,
+
+    pub contexts: Vec<MemoryContextState<F>>,
+}
+
+impl<F: Field> Default for MemoryState<F> {
+    fn default() -> Self {
+        Self {
+            log: vec![],
+            // We start with an initial context for the kernel.
+            contexts: vec![MemoryContextState::default()],
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+pub(crate) struct MemoryContextState<F: Field> {
+    /// The content of each memory segment.
+    pub segments: [MemorySegmentState<F>; NUM_SEGMENTS],
+}
+
+#[derive(Default, Debug)]
+pub(crate) struct MemorySegmentState<F: Field> {
+    pub content: Vec<[F; VALUE_LIMBS]>,
```

Comment:
Yeah it's a bit much 😅 but I think the access pattern should be reasonably cache-friendly. There should only be a few `MemorySegmentState`s that we're likely to interact with around the same time, like the current context's stack and memory (both should be a few kilobytes generally) and any frequently-used bits of kernel code.

I think we might end up with the property that all memory values fit in 256 bits; if so we could ~halve memory use by storing `U256`s instead of `[F; 8]`. I'm not certain though, so let's revisit a bit later.